### PR TITLE
Added test coverage and SpatialDropout2D for custom_unet

### LINKED ISSUE #2: Is development still occuring?
I think this package is great! Is it still under active development? If yes, my team and I would like to contribute, as we have recently started using it for our work.

### PR DESCRIPTION
**Pull request by gkaissis**
-Modified custom_unet to use Spatial Dropout
-Added a Changelog
-Bumped version to 0.0.8
-Removed nonsignificant whitespace
-Added tests
-Closes: #2

### CODE REVIEW

**Comment by karolzak** [comment]
Hi @gkaissis !
Thank you so much for your contribution!

Very interesting approach with `SpatialDropout` - that's the first time I hear about it so thank you for bringing that up! Looks like it indeed makes more sense to use it vs regular `Dropout` however since this is `custom_unet` I would like to leave it to users decision. I'm ok with setting `SpatialDropout` as the default but there should be an argument that would allow to change it back to regular easily.

Other than that I will look into the changes you made and make some comments wherever necessary so please address them when you have a moment.

 Also, before we decide to merge and bump version I need to fix requirements.txt and build+release pipelines to automate the process.

**Comment by gkaissis** [comment]
Dear @karolzak,

admittedly I opened this as a sort of preliminary PR before spending a lot of time on the PR and risking receiving no reply (kinda to test the waters since the project had been dormant for a while). I'd be happy to contribute more since I've done a lot of segmentation work over the past months in case you have the time and energy to maintain the project (alternatively, I'm happy to help out with maintenance as well). Me and my student might be able to contribute:
-3D UNets
-Attention gates
-Attention Gated Recurrent UNets
-Help out with documentation, notebooks and tests
etc.


Please let me know how you see the future of the project and I'll see what we can do to help out!

George

File: keras_unet/models/custom_unet.py
Comment (karolzak) [inline keras_unet/models/custom_unet.py:63]: @gkaissis can you share more context on why use_bias should be `False` when using `batch norm`?

File: keras_unet/models/custom_unet.py
Comment (karolzak) [inline keras_unet/models/custom_unet.py]: @gkaissis I understand that you wanted to move `Activation` layer after the `BatchNormalization` layer but in this case activation layer line shouldn't be placed under `if use_batch_norm` condition. We need to add activation regardless of `use_batch_norm`.
As for `Conv2D->BN->Activation` vs `Conv2D->Activation->BN` can you share some context or research on why exactly you favor the first one over the later? My personal research shows that it either doesn't make any difference or applying BN after activation (e.g. ReLu) makes more sense because we normalize output after removing negative values. Would love to get your perspective on that!

**Review by karolzak — CHANGES_REQUESTED** [review]
Did a quick review with just 2 comments. Please address them before we can merge.
Thanks for putting in some tests! Much appreciated!

**Comment by karolzak** [comment]
> Dear @karolzak,
> admittedly I opened this as a sort of preliminary PR before spending a lot of time on the PR and risking receiving no reply (kinda to test the waters since the project had been dormant for a while). I'd be happy to contribute more since I've done a lot of segmentation work over the past months in case you have the time and energy to maintain the project (alternatively, I'm happy to help out with maintenance as well). Me and my student might be able to contribute:
> -3D UNets
> -Attention gates
> -Attention Gated Recurrent UNets
> -Help out with documentation, notebooks and tests
> etc.
> Please let me know how you see the future of the project and I'll see what we can do to help out!
> George

This sounds great! I would love to see more of your contributions and I can promise to stay as active as possible here and maintain everything.
I have plenty of new stuff I'd like to add to this repo too.
And I think I will be able to commit more time to this project in near future.

As for the future of this project I'm pretty flexible to be honest. Anything that makes sense and could be useful in semantic segmentation is more than welcome in this repo!

File: keras_unet/models/custom_unet.py
Comment (gkaissis) [reply keras_unet/models/custom_unet.py]: Fully agree, this is an indentation issue.
There seems to be no absolute consensus whether BN before activation or after makes a difference (which as you say usually is a sign that it doesn't make a difference). 
That said, there are [some experiments](https://github.com/ducha-aiki/caffenet-benchmark/blob/master/batchnorm.md) stating slightly higher performance with activation after BN, but I don't believe it to be a reproducibly better choice. If you prefer, we can leave it in the layer specification.

>Simple is better than complex.

File: keras_unet/models/custom_unet.py
Comment (gkaissis) [reply keras_unet/models/custom_unet.py:63]: Batch Norm includes a bias parameter that just rescales the bias from the previous layer to whatever is needed. It's not problematic to use bias with batch norm, but it's unnecessary since the BN will do what it wants anyway and it reduces performance a little.

**Comment by gkaissis** [comment]
> This sounds great! I would love to see more of your contributions and I can promise to stay as active as possible here and maintain everything.
> I have plenty of new stuff I'd like to add to this repo too.
> And I think I will be able to commit more time to this project in near future.
> 
> As for the future of this project I'm pretty flexible to be honest. Anything that makes sense and could be useful in semantic segmentation is more than welcome in this repo!

Great! In this case, I'll devote some time and ask my group to help out as well if they have time!

File: keras_unet/models/custom_unet.py
Comment (karolzak) [reply keras_unet/models/custom_unet.py]: > That said, there are some **experiments stating slightly higher performance with activation after BN**...

Hmm checked the experiment you linked there and it seems like it actually [states the opposite](https://github.com/ducha-aiki/caffenet-benchmark/blob/master/batchnorm.md#bn----before-or-after-relu):

>BN -- before or after ReLU?
>
>Name | Accuracy | LogLoss | Comments
>-- | -- | -- | --
>Before | 0.474 | 2.35 | As in paper
>Before + scale&bias layer | 0.478 | 2.33 | As in paper
>After | 0.499 | 2.21 |  
>After + scale&bias layer | 0.493 | 2.24 |  
>
>**So in all next experiments, BN is put after non-linearity**

According to these experiments BN after non-linearity gives better results.
But nevertheless I agree with you that there seems to be no absolute consensus in this topic..
For now let's keep the activation inside layer specification as it was before

**Comment by karolzak** [comment]
> Great! In this case, I'll devote some time and ask my group to help out as well if they have time!

Thank you in advance @gkaissis and your students as well! :)

File: keras_unet/models/custom_unet.py
Comment (karolzak) [reply keras_unet/models/custom_unet.py:63]: Thanks for explaination! Makes perfect sense

File: keras_unet/models/custom_unet.py
Comment (gkaissis) [reply keras_unet/models/custom_unet.py]: Hmm, you are correct. Misread that... Apologies.

**Comment by gkaissis** [comment]
Addressed some of the issues raised. ReLu is now the default activation before the BN layer, dropout type is selectable, added a docstring and some new tests and left version at 0.0.7.

File: keras_unet/models/custom_unet.py
Comment (karolzak) [inline keras_unet/models/custom_unet.py]: Can you change it to `activation=activation,` so that we don't have it hardcoded?

**Review by karolzak — CHANGES_REQUESTED** [review]

File: keras_unet/models/custom_unet.py
Comment (gkaissis) [reply keras_unet/models/custom_unet.py]: Yeah sorry, fixed it

**Comment by gkaissis** [comment]
@karolzak are there any more changes required or can we merge? My colleague has another PR ready which he'd like to submit soon.
Best, G

**Review by karolzak — APPROVED** [review]

**Pull request merged**
